### PR TITLE
refactor: use raw_ptr for class fields that are pointers

### PR DESCRIPTION
**Pull request by ckerr**
#### Description of Change

This is a first stab at using upstream's `raw_ptr` mechanism as described in https://chromium.googlesource.com/chromium/src/+/main/base/memory/raw_ptr.md and https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md#non_owning-pointers-in-class-fields.

WfM locally, but I'm interested to see what happens when feeding this into CI. Marking as a draft pending a little CI feedback.

No particular stakeholders, but @electron/wg-releases and @electron/wg-security might be interested

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

**Review by RaisinTen — COMMENTED** [review]
Appears to be a bit more useful than `std::experimental::observer_ptr` because of the zero initialization and clearing on destruction/move bits but that seems to be available only on platforms where `USE_BACKUP_REF_PTR` is enabled and I haven't checked where that is enabled, did you get a chance to check?

Also, were you going to put something after `for` in the title `refactor: use raw_ptr for`?

File: shell/renderer/electron_render_frame_observer.h [hunk 1]
```diff
@@ -45,8 +46,8 @@ class ElectronRenderFrameObserver : public content::RenderFrameObserver {
                           const std::string& channel);
 
   bool has_delayed_node_initialization_ = false;
-  content::RenderFrame* render_frame_;
-  RendererClientBase* renderer_client_;
+  raw_ptr<content::RenderFrame> render_frame_;
```
[the remark this answers is not in the document]
Comment (deepak1556) [reply shell/renderer/electron_render_frame_observer.h:49]: renderer is ignored for now due to regression in browser benchmark https://chromium-review.googlesource.com/c/chromium/src/+/3184804 but there is not enough context to know if any raw_ptr usage in the render process is good/bad. Either we measure this change or we can follow the exclusions https://chromium.googlesource.com/chromium/src/+/main/tools/clang/rewrite_raw_ptr_fields/manual-paths-to-ignore.txt

**Comment by ckerr** [comment]
> Also, were you going to put something after `for` in the title `refactor: use raw_ptr for`?

LOL! I got called away as I was pushing 'submit' and meant to come back a minute later to edit it, but didn't quite make it :smile:

**Comment by ckerr** [comment]
> Appears to be a bit more useful than `std::experimental::observer_ptr` because of the zero initialization and clearing on destruction/move bits but that seems to be available only on platforms where `USE_BACKUP_REF_PTR` is enabled and I haven't checked where that is enabled, did you get a chance to check?

tl;dr: It looks like this is available on all the platforms that Electron cares about.

There is a lot of logic to decide this in https://chromium.googlesource.com/chromium/src/+/refs/heads/main/build_overrides/partition_alloc.gni and in https://chromium.googlesource.com/chromium/src/base/+/refs/heads/main/allocator/allocator.gni. If you want to follow the decision tree, start at the end of partition_alloc.gni with `enable_backup_ref_ptr_support_default = _is_brp_supported` and work backwards from there.

File: shell/browser/api/electron_api_browser_view.h [hunk 2]
```diff
@@ -81,7 +82,7 @@ class BrowserView : public gin::Wrappable<BrowserView>,
   v8::Local<v8::Value> GetWebContents(v8::Isolate*);
 
   v8::Global<v8::Value> web_contents_;
-  class WebContents* api_web_contents_ = nullptr;
+  raw_ptr<class WebContents> api_web_contents_ = nullptr;
```
[the remark this answers is not in the document]
Comment (ckerr) [reply shell/browser/api/electron_api_browser_view.h:85]: b67792b52f71d60aefb8e89c0d3ecd37ea473a2c explicitly assigns the new raw_ptrs to nullptr

File: shell/renderer/electron_render_frame_observer.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (RaisinTen) [reply shell/renderer/electron_render_frame_observer.h]: Is there a standard thing we use for benchmarking these kinds of stuff?

File: shell/browser/api/electron_api_browser_view.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (RaisinTen) [reply shell/browser/api/electron_api_browser_view.h:85]: Yea, that sounds good, resolving :+1:
